akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala

```diff
@@ -203,19 +213,19 @@ private[stream] final class GraphInterpreter(
 
   // Maintains the next event (and state) of the connection.
   // Technically the connection cannot be considered being in the state that is encoded here before the enqueued
-  // connection event has been processed. The inAvailable and outAvailable arrays usually protect access to this
+  // connection event has been processed. The inStates and outStates arrays usually protect access to this
   // field while it is in transient state.
   val connectionStates = Array.fill[Any](assembly.connectionCount)(Empty)
 
   // Indicates whether the input port is pullable. After pulling it becomes false
   // Be aware that when inAvailable goes to false outAvailable does not become true immediately, only after
   // the corresponding event in the queue has been processed
-  val inAvailable = Array.fill[Boolean](assembly.connectionCount)(true)
+  val inStates = Array.fill[PortState](assembly.connectionCount)(Available)
```

Comment:
The inStates and outStates arrays probably can be packed together in the same array, encoding the two sides as bit-fields of an Int:
 - lowest two bits are encoding the inStates state
 - next two bits are encoding the outStates state
 - additional flags can encode the PushCompleted and CompletedWithValue states instead of wrapper objects (probably one flag for "hasElement" so it can replace isInstanceOf[HasElement], and another flag if a pushcomplete is in flight)